crates/lib/src/fetchers/gh_crate_meta/hosting.rs

```diff
@@ -0,0 +1,40 @@
+use url::Url;
+
+use crate::errors::BinstallError;
+
+#[derive(Copy, Clone, Debug)]
+pub enum GitHostingServices {
+    GitHub,
+    GitLab,
+    BitBucket,
+    SourceForge,
+    Unknown,
+}
+impl GitHostingServices {
+    pub fn guess_git_hosting_services(repo: &Url) -> Result<Self, BinstallError> {
+        use GitHostingServices::*;
+
+        match repo.domain() {
+            Some(domain) if domain.starts_with("github") => Ok(GitHub),
+            Some(domain) if domain.starts_with("gitlab") => Ok(GitLab),
+            Some(domain) if domain == "bitbucket.org" => Ok(BitBucket),
+            Some(domain) if domain == "sourceforge.net" => Ok(SourceForge),
+            _ => Ok(Unknown),
+        }
+    }
+
+    pub fn get_default_pkg_url_template(self) -> Option<&'static [&'static str]> {
+        use GitHostingServices::*;
+
+        match self {
+            GitHub => Some(&[
+                "{ repo }/releases/download/v{ version }/{ name }-{ target }-v{ version }.{ archive-format }",
+                "{ repo }/releases/download/v{ version }/{ name }-{ target }.{ archive-format }",
```

Comment:
ooohh multiple formats! I hadn't thought of that. Maybe we can separate the filename from the URL/path and support these two forms (plus the also-common `name-version-target`) everywhere?

Reply:
> ooohh multiple formats! I hadn't thought of that. Maybe we can separate the filename from the URL/path and support these two forms (plus the also-common name-version-target) everywhere?

Yeah!
That's a good idea.

Perhaps I should also update `PkgMeta` to support multiple `pkg-url`s?

Reply:
> ooohh multiple formats! I hadn't thought of that. Maybe we can separate the filename from the URL/path and support these two forms (plus the also-common name-version-target) everywhere?

I've `name-target-version` form for GitLab and SourceForge.

Reply:
> ooohh multiple formats! I hadn't thought of that. Maybe we can separate the filename from the URL/path and support these two forms (plus the also-common name-version-target) everywhere?

I've added more templates, covering the release schema of `cargo-nextest`.

Reply:
> Perhaps I should also update PkgMeta to support multiple pkg-urls?

I would probably do this in another PR.

Reply:
> Perhaps I should also update PkgMeta to support multiple pkg-urls?

hmmm let's make that a separate issue/PR I think